net: l2: wifi: Add interface arg for all commands

krish2718 opened the pull request:
In order to support working with multiple VIFs (e.g., STA + AP) add interface as an option for all commands.

The interface can belong to same chipset or different (hypothetical).

Note: Most commands have no args, so, instead of adding gettopt to them with a single arg, I took a shortcut to parse interface in a central way, but in order to avoid issues with those commands that use gettopt, added `-i` option to them (but unused) and also didn't use getopt to parse the interface (it permutes the args). This was least invasive approach.

krish2718 committed:
net: l2: wifi: Add interface arg for all commands

In order to support working with multiple VIFs (e.g., STA + AP) add
interface as an option for all commands.

The interface can belong to same chipset or different (hypothetical).

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -147,6 +147,51 @@ struct wifi_ap_sta_node {
+	 */
+	for (int i = 1; i < argc; i++) {
+		if ((strcmp(argv[i], "-i") == 0 || strcmp(argv[i], "--iface") == 0) && (i + 1 < argc)) {
+			iface_index = atoi(argv[i + 1]);
```
Please use `shell_strtol()` and check the return values properly.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -147,6 +147,51 @@ struct wifi_ap_sta_node {
+
+	if (iface_index != -1) {
+		iface = net_if_get_by_index(iface_index);
+		if (!iface) {
```
Please do `if (iface == NULL) {` instead here and other parts of this PR.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -843,8 +886,9 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
-		{"ignore-broadcast-ssid", required_argument, 0, 'i'},
+		{"ignore-broadcast-ssid", required_argument, 0, 'g'},
 		{"ieee-80211r", no_argument, 0, 'R'},
+		{"iface", required_argument, 0, 'i'},
```
Shouldn't this be optional argument, if user does not give it we figure it out then?

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -873,7 +917,7 @@ static int __wifi_args_to_params(const struct shell *sh, size_t argc, char *argv
 	params->verify_peer_cert = false;
 
-	while ((opt = getopt_long(argc, argv, "s:p:k:e:w:b:c:m:t:a:B:K:S:T:A:V:I:P:i:Rh",
+	while ((opt = getopt_long(argc, argv, "s:p:k:e:w:b:c:m:t:a:B:K:S:T:A:V:I:P:g:Rh:i",
```
Should be `i:` instead

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1248,14 +1292,15 @@ static int wifi_scan_args_to_params(const struct shell *sh,
 		{"ssid", required_argument, 0, 's'},
 		{"max_bss", required_argument, 0, 'm'},
 		{"chans", required_argument, 0, 'c'},
+		{"iface", required_argument, 0, 'i'},
```
`optional_argument`?

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1363,9 +1411,6 @@ static int cmd_wifi_scan(const struct shell *sh, size_t argc, char *argv[])
 		if (opt_num < 0) {
 			shell_help(sh);
 			return -ENOEXEC;
-		} else if (!opt_num) {
-			PR_WARNING("No valid option(s) found\n");
-			do_scan = false;
```
Did not understand why we need to remove this

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -1921,17 +1966,18 @@ static int twt_args_to_params(const struct shell *sh, size_t argc, char *argv[],
 		{"info-disable", required_argument, 0, 'd'},
 		{"exponent", required_argument, 0, 'e'},
 		{"mantissa", required_argument, 0, 'm'},
+		{"iface", required_argument, 0, 'i'},
```
optional?

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2311,17 +2357,18 @@ static int wifi_ap_config_args_to_params(const struct shell *sh, size_t argc, ch
 		{"ht_capab", required_argument, 0, 'n'},
 		{"vht_capab", required_argument, 0, 'c'},
 #endif
+		{"iface", required_argument, 0, 'i'},
```
ditto

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2435,10 +2482,11 @@ static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
 	static const struct option long_options[] = {
 		{"force", no_argument, 0, 'f'},
 		{"verbose", no_argument, 0, 'v'},
+		{"iface", required_argument, 0, 'i'},
```
and here too

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2435,10 +2482,11 @@ static int cmd_wifi_reg_domain(const struct shell *sh, size_t argc,
 	};
 
-	while ((opt = getopt_long(argc, argv, "fv", long_options, &opt_index)) != -1) {
+	while ((opt = getopt_long(argc, argv, "fvi", long_options, &opt_index)) != -1) {
```
`i:`

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2804,6 +2853,7 @@ void parse_mode_args_to_params(const struct shell *sh, int argc,
 		{"ap", no_argument, 0, 'a'},
 		{"softap", no_argument, 0, 'k'},
 		{"get", no_argument, 0, 'g'},
+		{"iface", required_argument, 0, 'i'},
```
optional?

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2813,31 +2863,42 @@ void parse_mode_args_to_params(const struct shell *sh, int argc,
 			break;
 		case 'i':
 			mode->if_index = (uint8_t)atoi(state->optarg);
+			/* Don't count iface as it's common for both get and set */
```
We should fix also the atoi() here at some point (in separate commit or PR). In fact all the atoi calls in this file should be converted to `shell_strtol()` and the return values should be properly checked.

krish2718 replied:
Will raise a new PR

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2998,13 +3053,14 @@ void parse_filter_args_to_params(const struct shell *sh, int argc,
 		{"ctrl", no_argument, 0, 'c'},
 		{"data", no_argument, 0, 'd'},
 		{"get", no_argument, 0, 'g'},
+		{"iface", required_argument, 0, 'i'},
```
You have the same change twice here

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -3671,76 +3728,86 @@ static int cmd_wifi_pmksa_flush(const struct shell *sh, size_t argc, char *argv[
-		      cmd_wifi_ap_enable, 2, 45),
-	SHELL_CMD_ARG(stations, NULL, "List stations connected to the AP", cmd_wifi_ap_stations, 1,
-		      0),
+			  "-i --iface=<interface index>\n"
```
If there is only one interface, then specifying one should be optional

krish2718 replied:
Yes, that is how it works, it's always an optional arg, I can fix if its not in few places.

jukkar reviewed (commented):
No objections to this change, left some minor comment

krish2718 committed:
modules: hostap: Fix build

Due to recent POSIX rejig, we need to select an extra config option for
"gettimeofday".

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2900,7 +2968,7 @@ void parse_channel_args_to_params(const struct shell *sh, int argc,
 	struct getopt_state *state;
 	static const struct option long_options[] = {
-		{"if-index", optional_argument, 0, 'i'},
+		{"iface", optional_argument, 0, 'i'},
```
Shouldn't this be `required_argument`? At least the help text indicates that the interface value should be given.

jukkar commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -2998,7 +3066,7 @@ void parse_filter_args_to_params(const struct shell *sh, int argc,
 	struct getopt_state *state;
 	static const struct option long_options[] = {
-		{"if-index", optional_argument, 0, 'i'},
+		{"iface", optional_argument, 0, 'i'},
```
ditto

jukkar reviewed (commented):
Looks ok. As this is changing the cmd options that user might use, please add a note to migration guide for this.

krish2718 committed:
doc: releases: Update migration with Wi-Fi shell

Document old and new options for Wi-Fi shell to help with the migration.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

jukkar reviewed (commented):
The changes look ok, minor nit about code indentation in various shell macros

krish2718 commented on subsys/net/l2/wifi/wifi_shell.c:
```diff
@@ -3671,77 +3739,87 @@ static int cmd_wifi_pmksa_flush(const struct shell *sh, size_t argc, char *argv[
 SHELL_STATIC_SUBCMD_SET_CREATE(
 	wifi_cmd_ap,
-	SHELL_CMD_ARG(disable, NULL, "Disable Access Point mode.\n", cmd_wifi_ap_disable, 1, 0),
+	SHELL_CMD_ARG(disable, NULL, "Disable Access Point mode.\n"
```
Removed all whitespace/indent changes. If we had https://github.com/zephyrproject-rtos/zephyr/pull/77546 then it would solve all such issues :)
